### PR TITLE
Create subgroup with google or facebook

### LINKED ISSUE #666: create subgroup w/ FB or Google OAuth [2]
resolved by #678 
_____________________________

# Acceptance Criteria

GIVEN I am a logged-out user
AND I have a FACEBOOK account with an email not tied to an Affinity account
AND I am visiting the subgroup creation page for Group Foo
* THEN I will see a "Create with Facebook" button
* WHEN I click the "Create with Facebook" button
* THEN I will be forwarded to the `members/new` page to gather my zip and phone number
* WHEN I fill out the form successfully
* THEN a new group will be created, I will be made its organizer, and I will receive a welcome email
* AND THEN I will be forwarded to the group's dashboard page

GIVEN I am a logged-out user
AND I have a FACEBOOK account with an email tied to an existing Affinity account
AND I **do not have a zip code** on record with Affinity
AND I am visiting the subgroup creation page for Group Foo
* THEN I will see a "Create with Facebook" button
* WHEN I click the "Create with Facebook" button
* THEN I will be forwarded to the `members/new` page to gather my zip and phone number
* WHEN I fill out the form successfully
* THEN a new group will be created, I will be made its organizer, and I will receive a welcome email
* AND THEN I will be forwarded to the group's dashboard page

GIVEN I am a logged-out user
AND I have a FACEBOOK account with an email tied to an existing Affinity account
AND I **do have a zip code** on record with Affinity
AND I am visiting the subgroup creation page for Group Foo
* THEN I will see a "Create with Facebook" button
* WHEN I click the "Create with Facebook" button
* THEN a new group will be created, I will be made its organizer, and I will receive a welcome email
* AND THEN I will be forwarded to the group's dashboard page

GIVEN I have made an error in any of the forms above
* THEN I will be prompted to correct them and not allowed to proceed until I do

GIVEN I have a GOOGLE account, all of the flows described above will work as described for FACEBOOK

### PR DESCRIPTION
**Pull request by vaughanj10**
Closes #666

### CODE REVIEW

File: app/controllers/people/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -55,17 +72,29 @@ def signup
   ###########
 
   def handle_new_person
-    redirect_to new_group_member_path(signup_mode: @service,
-                                      group_id: current_group.id,
-                                      person: { oauth: encrypt_token(@auth) })
+    case @signup_reason
```
Comment (aguestuser) [inline app/controllers/people/omniauth_callbacks_controller.rb]: for later: this is starting to get pretty deeply branched. if it keeps getting that way, let's consider introducing a helper concern something like: https://gist.github.com/aguestuser/541ef99a90aa405036da21a7e6fd8092 ?

File: app/controllers/subgroups_controller.rb [hunk 2]
```diff
@@ -3,25 +3,54 @@ class SubgroupsController < ApplicationController
...
+        omniauth_value,
+        subgroup_json: JSON.generate(subgroup_params.to_h),
+        auth_action:   'signup',
+        signup_reason: 'create_group',
+        group_id: @group.id
+      )
+    end
+  end
+
+  # GET groups/:group_id/subgroups/oauth_signup
```
Comment (aguestuser) [inline app/controllers/subgroups_controller.rb]: maybe one more comment to make clear why we do this? (so that we don't GET and POST to same endpoint?)

File: app/controllers/subgroups_controller.rb [hunk 3]
```diff
@@ -78,28 +144,29 @@ def organizer_params
...
-              personal_addresses_attributes: [:postal_code])
+              personal_addresses_attributes: [:postal_code]
+             )
        .tap { |prams| format(prams) }
   end
 
-  def format(prams)
-    prams
+  def format(params)
+    params
```
Comment (aguestuser) [inline app/controllers/subgroups_controller.rb]: hmm... i'd actually prefer `prams` here. `params` shadows a variable that is already in scope for the controller. which is why i used `prams`. fine with some other convention, but whatever we call it, i'd offer that it introduces unhelpful confusion to call it `params`

File: app/controllers/subgroups_controller.rb [hunk 4]
```diff
@@ -114,11 +181,33 @@ def handle_create_success
 
   def handle_create_error
     if @signup_mode
-      @person = @subgroup.memberships.first.person
+      @errors = AggregateError.new(objects: [@person, @subgroup]).errors
```
Comment (aguestuser) [inline app/controllers/subgroups_controller.rb]: cool!

File: app/services/aggregate_error.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class AggregateError
```
Comment (aguestuser) [inline app/services/aggregate_error.rb]: nice!

File: app/services/oauth.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+class Oauth
```
Comment (aguestuser) [inline app/services/oauth.rb]: nice extraction!

File: test/feature/create_group_test.rb [hunk 7]
```diff
@@ -45,306 +47,417 @@ class CreateGroupTest < FeatureTest
...
           end
 
-          it "has correct required fields" do
-            ['Email*',
-             'Password*',
-             'First Name*',
-             'Last Name*',
-             'Zip Code*'].each do |label|
-              "input[placeholder='#{label}'][required='required']"
+          describe "via facebook signup" do
```
Comment (aguestuser) [inline test/feature/create_group_test.rb]: just noting: there is no browser test for google oauth first-time user signup.

i understand that the facebook cases cover almost all of the same logic. however, given how hair the whole thing is (and thus how likely to break for reasons that might elude us later), i'd argue that at least one path for the happy path on the google oauth path might be warranted? (i *think* i did this for the join group tests?) what do you think?

**Comment by vaughanj10** [comment]
# Notes for Manual Testing

To verify this works for all corner cases, do the following:

## Setup

1. Login to both Facebook and Gmail with testy@affinity.works
2. In the console, use the class method `Person.find_by_email("testy@affinity.works")` to ensure that no user with that email already exists

##  I. Testing New User Path

A. For Google:

1. go to `/groups/2/subgroups/new`
2. click "Create with Google"
3. verify that you are logged in and forwarded to second sign up screen `/groups/2/subgroups/oauth_signup`
4. Enter in zip code and phone number
5. Submit
6. Verify that you are redirected to the new subgroups dashboard
7. Log out
8. in console, use `Person.find_by_email("testy@affinity.works").destroy` to destroy testy

B. For Facebook:


1. go to `/groups/2/subgroups/new`
2. click "Create with Google"
3. verify that you are logged in and forwarded to second sign up screen `/groups/2/subgroups/oauth_signup`
4. Enter in zip code and phone number
5. Submit
6. Verify that you are redirected to the new subgroups dashboard
7. Log out
8. in console, use `Person.find_by_email("testy@affinity.works").destroy` to destroy testy

File: app/controllers/subgroups_controller.rb (hunk 3 above)
Comment (aguestuser) [reply app/controllers/subgroups_controller.rb]: (could call it `_params` if the mis-spelling also introduces unnecessary confusion?)

File: test/feature/create_group_test.rb (hunk 7 above)
Comment (aguestuser) [reply test/feature/create_group_test.rb]: and realizing now: no tests to cover facebook (or google) signup for already-existing user (who is or isn't missing contact info)

File: app/controllers/subgroups_controller.rb [hunk 8]
```diff
@@ -76,30 +159,31 @@ def organizer_params
...
-    prams
+  def format(_params)
+    _params
       .tap { |p| format_contact_info(p) }
       .tap { |p| handle_empty_contact_info(p) }
   end
 
-  def format_contact_info(prams)
-    return if current_person # only gather contact info for logged-out users
+  def format_contact_info(params)
```
Comment (vaughanj10) [inline app/controllers/subgroups_controller.rb]: ^-

File: app/controllers/subgroups_controller.rb [hunk 9]
```diff
@@ -76,30 +159,31 @@ def organizer_params
...
+      'phone_numbers_attributes',  'email_addresses_attributes',  'personal_addresses_attributes'
+    ].each do |collection| 
+      params.dig(collection, '0').merge!(primary: true) if params.dig(collection, '0')
+    end.compact
   end
 
-  def handle_empty_contact_info(prams)
-    if prams.dig('phone_numbers_attributes', '0', 'number')&.empty?
-      prams.delete('phone_numbers_attributes')
+  def handle_empty_contact_info(params)
```
Comment (vaughanj10) [inline app/controllers/subgroups_controller.rb]: ^-

File: app/views/shared/_signup_aggregate_error_alert.html.erb [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<%# locals: person: Person %>
```
Comment (vaughanj10) [inline app/views/shared/_signup_aggregate_error_alert.html.erb]: REMOVE

**Comment by aguestuser** [comment]
okay! i have run through all the QA steps on a staging review app and verified that all features work as designed. 

i am going to merge, ship, and mark as "NEEDS REFACTOR" and refrain from moving to "done" until we have completed some/all of the below priorities @vaughanj10 identified during code-review yesterday:

* cleanup before_action order dependency in member_controller
* extract shared logic btw/ members_controller, subgroups_controller (particularly w.r.t. serializing/deserializing/parsing oauth params)
* move oauth signup handler logic to concern or helper class
** use classes for MemberState w/ uniform `handle` methods
** separate out controller dispatch logic (member/signup) from member state classification (do the former first where possible -- ie: in create group flow)
* eliminate red herring `primary: true` dead code
* make sure shared code for oauth partials/layouts is used in all cases
* (probably a few more things i'm forgetting)
